### ISSUE TITLE
Evaluatin a path immediately triggers any function and variables in it

### ISSUE THREAD

**Issue by ZeAntwan** [astrochili/narrator#21]
Hi, I know I might be digging back this library, but I'm happy to try and explore solutions for this!

I've been working on porting this library to work on Playdate (meaning needed some simplification), but I'm also trying to get it to evaluate ink knots/paragraph line by line. While I can get the text to work as expected using the continue(n) function, the problem I have is that any variable that get set in it or any functions that get called, are triggered when the path is jumped to.

This can be tested using the example code and the readme, and calling this "call_911" knot
```
EXTERNAL phonebook(name,number)
EXTERNAL test(pouet)

=== function phonebook(name,number) ===
// Placeholder for phonebook function
~ return 1

=== function test(pouet) ===
// Placeholder for phonebook function
~ return 1

=== call_911 ===
{call_911 >= 2: I've seen you before... }
You've just called the emergency services
<i>Test</i>
~ test("test")
~ phonebook("Emergency","922")
[i]Test2[/i]
Please don't
We have shit to do 
And this is a very long thing
It should remove some lines from before
Let's hope so
-> DONE
```

**Comment by astrochili** [comment]
Hi! 

As I understand it, you mean that after calling `story:begin()` or `story:choose(answer)` there is a complete internal paragraphs executins with all external functions called, and then the host game can pop paragraphs from the ready-to-read stack. So there is no way to place an external call between paragraph output. Am I right?

**Comment by ZeAntwan** [comment]
Yes, I think that matches the issue I'm having. So far, I've only tried using `story:continue()` and was expecting that doing `story:continue(1)` would only process one line, and not the whole paragraph before returning a line. I haven't tried choose to see if something similar happen (but would expect it to after looking at the code and my understanding of it).

And I will double check, but I believe this applies to any kind of variable operation too, so setting a value anywhere in the paragraph, will set the value when the knot is jumped to instead of wait for it to be "read".

In my example, the "phonebook" function should add something only once the line is read, which seems to be the behaviour of most other ink libraries I'm using for C# (but I could be wrong)

**Comment by astrochili** [comment]
That's definitely how it works now. The `continue(x)` function just returns already prepared paragraphs.

If it doesn't work that way in the other frameworks, it's probably a Narrator design bug 😔.

You can also check out this implementation that works with compiled js files - https://github.com/abadonna/defold-ink/tree/main/ink

**Comment by astrochili** [comment]
We've had a little chat with @abadonna about this and to be honest, we're not sure if other plugins do things differently.

Yes, it would have to be tested to make that claim. But logical reasoning about the liquid nature of paragraphs, when often without code execution it is not clear where the paragraphs are and where they are not, up to the fact that functions themselves can return a whole set of paragraphs... give the impression that the current behaviour is correct.

And for your idea, indeed, tags are suitable. Or splitting into blocks by an additional intermediate choice, which is "especially" automatically handled in the game. The second way seems to be the most architecturally correct for your example.

**Comment by abadonna** [comment]
Just checked the original Unity asset. 
If **ContinueMaximally()** method is called - external function called first, same behaviour as in Narrator.
If **Continue()** method is called and only 1 paragraph is returned - external function called as @ZeAntwan wants it to be called ... 
It's pretty straightforward, story starts to read ink script, and if it need to return result - story just stops, so in case of single paragraph - it just returned and in case of multiple - every function calls processed and paragraphs are generating until story finally finds the choice.

**Comment by astrochili** [comment]
Thanks for checking that out.

That's very interesting, sounds like not an obvious difference in behaviour. It turns out if calling continue(1) 3 times is not the same as calling contnue(3) 🤯

On the other hand, Ink has always had room for ambiguity 🙃

**Comment by ZeAntwan** [comment]
Thanks for looking into it, and sorry I did not provide the test. I'm using Godot, which uses a port of the Unity/C# implementation of it and that might be what threw me off.

It's always been a fight for me to understand how to "properly" use ink when trying to do adventure or dialogue box bas games that show dialogues line by line.

With that said, like you mentioned, while I still need to go a bit deeper, I might be able to manage using tags for now 🙂
